Build a storage reader/writer to write checkpoints in HF format

ankitageorge opened the pull request:
Summary: Title - we want to write checkpoints in HF format with DCP, this diff allows this for the non-distributed use case.

Test Plan:
buck2 test 'fbcode//mode/dev-nosan' fbcode//caffe2/test/distributed/checkpoint:test_hf_torchtune_storage

N6476188 --> able to save and load tensor in hf format

Differential Revision: D68444967




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0

wconstab commented:
Cc @tianyu-l @fegin

fegin commented:
The only concern is that if HF changes the checkpointing format or invents a new format, then PyTorch team has to fill the gap. cc., @pradeepfn

saumishr commented on torch/distributed/checkpoint/hf_torchtune_storage.py:
```diff
@@ -0,0 +1,277 @@
+
+__all__ = ["HuggingFaceHubTorchTuneWriter", "HuggingFaceHubTorchTuneReader"]
+
+_metadata_fn: str = "model.safetensors.index.json"
+
+FILE_NAME = "model-{cpt_idx}-of-{num_shards}"
+NUM_ITEMS_PER_FILE = 15
+SUFFIX = ".safetensors"
```
This logic is TorchTune specific and doesn't belong in DCP. It should be moved to TorchTune and probably can be passed down to the reader/writer constructor.

ankitageorge replied:
I think all HF checkpoints have this format and don't think this is TorchTune specific

saumishr commented on torch/distributed/checkpoint/hf_torchtune_storage.py:
```diff
@@ -0,0 +1,277 @@
+        plan: SavePlan,
+        planner: SavePlanner,
+    ) -> Future[List[WriteResult]]:
+        storage_plan : Dict[str, int] = plan.storage_data
```
If the plan objects are ready, why do we need to customize the write_data method. Shouldn't the base class's method be used: https://github.com/pytorch/pytorch/blob/f35e60b21c85718ade2d547fe00e42af7cf237b9/torch/distributed/checkpoint/filesystem.py#L582 ?

saumishr commented on torch/distributed/checkpoint/hf_torchtune_storage.py:
```diff
@@ -0,0 +1,277 @@
+        storage_md = {}
+        for wr_list in results:
+            storage_md.update({wr.index.fqn: wr.storage_data for wr in wr_list})
+        metadata_to_write["weight_map"] = storage_md
```
This weight_map contains the mapping of state_dict FQNs to checkpoint files? Is it for the distributed case when we may have to locate which files get loaded based on the state dict in that rank?

ankitageorge replied:
Yes. I also was just copying the existing metadata structure https://huggingface.co/meta-llama/Llama-3.1-8B-Instruct/blob/main/model.safetensors.index.json

ankitageorge replied:
btw I reached out Joe about this and the reason they have many files even for the single rank case is because HFHub doesn't like having more than 5gb for safetensors files

saumishr commented on torch/distributed/checkpoint/hf_torchtune_storage.py:
```diff
@@ -0,0 +1,277 @@
+        return _metadata_fn
+
+
+class HuggingFaceHubTorchTuneReader(FsspecReader):
```
I don't see any registration of the HFFileSystem. I'm not sure how the files with HF remote path are working. Are you only handing the case of reading/writing HF files from local filesystem?

ankitageorge replied:
Ya I'm not sure how it works without registering, but it is working and writing to hf. You can verify with N6527225

ankitageorge commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -456,13 +456,17 @@ class FileSystem(FileSystemBase):
-        with cast(Path, path).open(mode) as stream:
+        if not isinstance(path, Path):
+            path = Path(path)
+        with path.open(mode) as stream:
```
I made a separate PR for this https://github.com/pytorch/pytorch/pull/145751, but you can't stack in github, so had to include it here so that the test I wrote passes.

fegin commented on torch/distributed/checkpoint/hf_torchtune_storage.py:
```diff
@@ -0,0 +1,277 @@
+        fut.set_result(None)
+        return fut
+
+    def read_metadata(self) -> Metadata:
```
Do we have an unittest for this method? I think reading is very important as many users mostly want to read from HF checkpoints and import the checkpoint to TorchTitan/TorchTune with DCP. The current pain point is that users will need to iterate the `weight_map` and the index files and do the conversion.

ankitageorge replied:
will add one

saumishr commented on torch/distributed/checkpoint/_hf_storage.py:
```diff
@@ -0,0 +1,192 @@
+        from  huggingface_hub import HfFileSystem
+
+        if HfFileSystem.protocol not in fsspec.available_protocols():
+            fsspec.register_implementation(HfFileSystem.protocol, HfFileSystem)
```
Please check if multiple registrations can cause any issues if this object gets initialized on every save request.

ankitageorge replied:
we're only registering if the customer hasn't registered so we wouldn't run into this right? or should we just skip the registering all together and leave it to the users since we're already leaving the huggingface instillation to users?

saumishr commented on torch/distributed/checkpoint/_hf_storage.py:
```diff
@@ -0,0 +1,192 @@
+        super().prepare_local_plan(plan)
+        return dataclasses.replace(plan, storage_data=self._fqn_to_index_mapping)
+
+
```
Lint: Extra empty line

saumishr commented on torch/distributed/checkpoint/__init__.py:
```diff
@@ -2,6 +2,7 @@
 from .api import CheckpointException
 from .default_planner import DefaultLoadPlanner, DefaultSavePlanner
 from .filesystem import FileSystemReader, FileSystemWriter
+from ._hf_storage import HuggingFaceHubReader, HuggingFaceHubWriter
```
`nit` We can consider renaming these to `HuggingFaceStorageReader` and `HuggingFaceStorageWriter`. This way it would be easier to identify all the storage reader/writer components supported in DCP.

saumishr reviewed (commented):
Looks great overall. Only a few minor comments. Awesome work!

saumishr commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -303,6 +308,7 @@ def _write_item(
     data: Union[io.BytesIO, torch.Tensor],
     write_item: WriteItem,
     storage_key: str,
+    safe_tensors: bool = False,
```
For the first iteration, its okay to change the internal methods. However, the longer term plan was to provide a serializer interface and do a dependency injection to the default storage reader/writer components. For HuggingFace, we can inject the SafeTensors based serializer. Not a blocker though. Lets iterate on it a bit in follow up PRs to provide the extensibility here instead of forcing HuggingFaceStorage reader/writer users to use the SafeTensors.

saumishr reviewed (approved):
Looks great. Please fix the lint errors before merging. Lets follow up on how to make serializer more extensible.

ankitageorge committed:
Build a storage reader/writer to write checkpoints in HF format (#146352)

Summary:

Title - we want to write checkpoints in HF format with DCP, this diff allows this for the non-distributed use case

Test Plan:
buck2 test 'fbcode//mode/dev-nosan' fbcode//caffe2/test/distributed/checkpoint:test_hf_torchtune_storage

N6596613 --> able to save and load tensor in hf format

Reviewed By: saumishr

Differential Revision: D68444967

jeanschmidt commented:
@pytorchbot revert -m "Author ignored linting errors" -c ignoredsignal

jeanschmidt commented:
Closing this PR as the original diff is reverted internally
